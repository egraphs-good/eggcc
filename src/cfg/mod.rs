//! Parse a bril program into a CFG.
//!
//! The methods here largely ignore the instructions in the program: all that we
//! look for here are instructions that may break up basic blocks (`jmp`, `br`,
//! `ret`), and labels. All other instructions are copied into the CFG.
use core::fmt::Debug;
use std::fmt::Formatter;
use std::str::FromStr;
use std::{collections::HashMap, fmt::Display};
use std::{fmt, mem};

use bril_rs::{Argument, Code, EffectOps, Function, Instruction, Position, Program, Type};
use petgraph::dot::Dot;

use petgraph::stable_graph::{EdgeReference, StableDiGraph};
use petgraph::visit::{EdgeRef, Visitable};
use petgraph::{
    graph::NodeIndex,
    visit::{DfsPostOrder, Walker},
};

use crate::rvsdg::from_cfg::FunctionTypes;
use crate::util::{run_cmd_line, ListDisplay, Visualization};

/// A subset of nodes for a particular CFG.
pub(crate) type NodeSet = <StableDiGraph<BasicBlock, Branch> as Visitable>::Map;

#[cfg(test)]
pub(crate) mod tests;

pub(crate) mod structured;
pub(crate) mod to_bril;
pub(crate) mod to_structured;

/// Convert a program to a cfg.
/// Loops over all the functions, translating individually.
pub(crate) fn program_to_cfg(program: &Program) -> SimpleCfgProgram {
    let mut functions = Vec::new();
    for func in &program.functions {
        let cfg = function_to_cfg(func);
        functions.push(cfg);
    }
    CfgProgram { functions }
}

#[derive(Clone)]
pub struct CfgProgram<CfgType> {
    pub functions: Vec<CfgFunction<CfgType>>,
}

/// Simple programs only branch on booleans defined in bril
pub type SimpleCfgProgram = CfgProgram<Simple>;
/// Switch programs can also branch on values defined in annotations
pub type SwitchCfgProgram = CfgProgram<Switch>;

impl SimpleCfgProgram {
    /// Convert a simple program to a switch program
    /// trivial, since simple programs are a subset of switch programs
    pub fn into_switch(self) -> SwitchCfgProgram {
        SwitchCfgProgram {
            functions: self
                .functions
                .into_iter()
                .map(|f| f.into_switch())
                .collect(),
        }
    }
}

impl SimpleCfgFunction {
    pub fn into_switch(self) -> SwitchCfgFunction {
        SwitchCfgFunction {
            args: self.args,
            graph: self.graph,
            entry: self.entry,
            exit: self.exit,
            name: self.name,
            return_ty: self.return_ty,
            _phantom: Switch,
        }
    }
}

impl<CfgType> CfgProgram<CfgType> {
    pub(crate) fn function_types(&self) -> FunctionTypes {
        let mut types = FunctionTypes::default();
        for func in &self.functions {
            let output_type = func.return_ty.clone();
            types.insert(func.name.clone(), output_type);
        }
        types
    }

    pub(crate) fn visualizations(&self) -> Vec<Visualization> {
        let mut visualizations = vec![];
        for function in &self.functions {
            let svg = function.to_svg();
            visualizations.push(Visualization {
                result: svg,
                file_extension: ".svg".to_string(),
                name: function.name.clone(),
            });
        }
        visualizations
    }
}

/// The name (or label) associated with a basic block.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum BlockName {
    /// The unique entrypoint for a function.
    Entry,
    /// The unique exit point for a function (assuming the function is not an infinite loop).
    Exit,
    /// An unnamed block generated as part of the restructuring process.
    Placeholder(usize),
    /// A named block from the original Bril program.
    Named(String),
}

/// The distinguished identifier associated with the return value of a function,
/// if it has one.
pub(crate) fn ret_id() -> Identifier {
    Identifier::Num(usize::MAX - 1)
}

/// The distinguished identifier associated with "state".
///
/// To recover implicit ordering dependencies between impure operations in bril,
/// we treat effectful operations as taking an "extra argument" and then
/// assigning to that a state variable. This allows the rest of the
/// transformation to preserve the ordering information. So code like:
///
/// > x: int = const 1
/// > print x
/// > y: int = const 2
/// > print y
///
/// Is effectively translated to:
///
/// > x: int = const 1
/// > <state> = print x <state>
/// > y: int = const 2
/// > <state> = print y <state>
///
/// `state_id` is the identifier corresponding to this <state> variable.
pub(crate) fn state_id() -> Identifier {
    Identifier::Num(usize::MAX)
}

impl Display for BlockName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            BlockName::Entry => write!(f, "entry___"),
            BlockName::Exit => write!(f, "exit___"),
            BlockName::Placeholder(n) => write!(f, "__{n}__"),
            BlockName::Named(s) => write!(f, "{}", s),
        }
    }
}

/// An number (`val`) between 0 and `of` (exclusive).
///
/// These are used to represent "switch" blocks in the CFG.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub(crate) struct CondVal {
    pub(crate) val: u32,
    pub(crate) of: u32,
}

impl From<bool> for CondVal {
    fn from(value: bool) -> Self {
        if value {
            CondVal { val: 1, of: 2 }
        } else {
            CondVal { val: 0, of: 2 }
        }
    }
}

impl FromStr for BlockName {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "entry___" => Ok(BlockName::Entry),
            "exit___" => Ok(BlockName::Exit),
            s => Ok(BlockName::Named(s.to_string())),
        }
    }
}

/// Identifiers either come from the source Bril program or are synthesized as
/// part of the RVSDG conversion process. The `Identifier` type stores both
/// kinds of name.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum Identifier {
    Name(String),
    Num(usize),
}

impl<T: AsRef<str>> From<T> for Identifier {
    fn from(value: T) -> Identifier {
        Identifier::Name(value.as_ref().into())
    }
}

impl fmt::Display for Identifier {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Identifier::Name(n) => {
                write!(f, "{n}")
            }
            Identifier::Num(n) => {
                write!(f, "@{n}")
            }
        }
    }
}

/// An annotation appended to the end of a basic block.
#[derive(Debug, PartialEq, Clone)]
pub(crate) enum Annotation {
    // not present in Simple CFGs
    AssignCond { dst: Identifier, cond: u32 },
    // can be present in Simple CFGs
    AssignRet { src: Identifier },
}

/// A branch-free sequence of instructions.
#[derive(Clone, PartialEq)]
pub struct BasicBlock {
    /// The primary instructions for a block.
    pub(crate) instrs: Vec<Instruction>,
    /// Any annotations added to the end of the block during restructuring.
    pub(crate) footer: Vec<Annotation>,
    /// The name for the block.
    pub(crate) name: BlockName,
    pub(crate) pos: Option<Position>,
}

impl Debug for BasicBlock {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut code_block = self.to_code();
        code_block.extend(self.debug_code_for_footer());
        write!(f, "{}", ListDisplay(code_block, "\n"))
    }
}

impl BasicBlock {
    pub(crate) fn empty(name: BlockName) -> BasicBlock {
        BasicBlock {
            instrs: Default::default(),
            footer: Default::default(),
            name,
            pos: None,
        }
    }

    /// Annotations should not be directly translated back into bril, as they
    /// include values (like "conditions" branching >2 ways) that do not have a
    /// clear bril type.
    ///
    /// This method is used for debug printing purposes only, and for that
    /// reason it "reflects" conditions back into bril to the best of its
    /// ability.
    fn debug_code_for_footer(&self) -> impl Iterator<Item = Code> + '_ {
        self.footer.iter().map(|ann| match ann {
            // Conditions are untyped and they do not generally make sense to directly "reflect" back into bril. For the purposes of printing.
            Annotation::AssignCond { dst, cond } => Code::Instruction(Instruction::Constant {
                dest: format!("{dst}"),
                op: bril_rs::ConstOps::Const,
                pos: None,
                const_type: if *cond < 2 { Type::Bool } else { Type::Int },
                value: match cond {
                    0 => bril_rs::Literal::Bool(false),
                    1 => bril_rs::Literal::Bool(true),
                    n => bril_rs::Literal::Int(*n as i64),
                },
            }),
            Annotation::AssignRet { src } => Code::Instruction(Instruction::Effect {
                op: EffectOps::Return,
                args: vec![format!("{src}")],
                funcs: vec![],
                labels: vec![],
                pos: None,
            }),
        })
    }

    fn to_code(&self) -> Vec<Code> {
        let mut instrs = Vec::new();
        instrs.push(Code::Label {
            label: self.name.to_string(),
            pos: self.pos.clone(),
        });
        instrs.extend(self.instrs.iter().map(|i| Code::Instruction(i.clone())));
        instrs
    }
}

/// A branch in the CFG.
#[derive(Debug, Clone)]
pub(crate) struct Branch {
    /// The type of branch.
    pub(crate) op: BranchOp,
    /// The position of the branch in the original program.
    #[allow(unused)]
    pub(crate) pos: Option<Position>,
}

/// The types of branch.
#[derive(PartialEq, Eq, Debug, Clone)]
pub(crate) enum BranchOp {
    /// An unconditional branch to a block.
    Jmp,
    /// A conditional branch to a block.
    Cond { arg: Identifier, val: CondVal },
}

#[derive(Debug, Clone)]
pub struct Switch;
#[derive(Debug, Clone)]
pub struct Simple;

/// The control-flow graph for a single function.
#[derive(Debug, Clone)]
pub struct CfgFunction<CfgType> {
    /// The arguments to the function.
    pub(crate) args: Vec<Argument>,
    /// The graph itself.
    pub(crate) graph: StableDiGraph<BasicBlock, Branch>,
    /// The entry node for the CFG.
    pub(crate) entry: NodeIndex,
    /// The (single) exit node for the CFG.
    pub(crate) exit: NodeIndex,
    /// The name of the function.
    pub(crate) name: String,
    pub(crate) _phantom: CfgType,
    pub(crate) return_ty: Option<Type>,
}

/// A simple CFG branches only on booleans, instead of allowing annotations
/// and branching on these annotations.
pub type SimpleCfgFunction = CfgFunction<Simple>;
pub type SwitchCfgFunction = CfgFunction<Switch>;

impl<CfgType> CfgFunction<CfgType> {
    pub(crate) fn has_return_value(&self) -> bool {
        self.return_ty.is_some()
    }

    pub fn to_dot(&self) -> String {
        format!("{:?}", Dot::new(&self.graph))
    }

    pub fn to_svg(&self) -> String {
        let dot_code = self.to_dot();
        run_cmd_line("dot", ["-Tsvg"], &dot_code).unwrap()
    }

    fn reverse_postorder(self: &CfgFunction<CfgType>) -> HashMap<BlockName, usize> {
        let mut reverse_postorder = HashMap::<BlockName, usize>::new();
        let mut post_counter = 0;
        DfsPostOrder::new(&self.graph, self.entry)
            .iter(&self.graph)
            .for_each(|node| {
                reverse_postorder.insert(self.graph[node].name.clone(), post_counter);
                post_counter += 1;
            });

        reverse_postorder
    }
}

#[derive(Debug, Clone)]
pub enum SimpleBranch {
    NoBranch, // must be the end of the function
    Jmp(BlockName),
    If {
        // arg is an integer specifying which branch to jump to
        arg: Identifier,
        then_branch: BlockName,
        else_branch: BlockName,
    },
}

impl SimpleCfgFunction {
    pub fn get_branch(&self, node: NodeIndex) -> SimpleBranch {
        let outgoing = self.graph.edges(node);
        match &outgoing.collect::<Vec<EdgeReference<Branch>>>().as_slice() {
            [] => {
                assert!(self.exit == node);
                SimpleBranch::NoBranch
            }
            [edge] => {
                let target: NodeIndex = edge.target();
                let branch = edge.weight();
                let BranchOp::Jmp = branch.op else {
                    panic!("Unexpected branch type");
                };
                SimpleBranch::Jmp(self.graph[target].name.clone())
            }
            [edge1, edge2] => match (&edge1.weight().op, &edge2.weight().op) {
                (
                    BranchOp::Cond {
                        arg: cond1,
                        val: CondVal { val: val1, of: 2 },
                    },
                    BranchOp::Cond {
                        arg: cond2,
                        val: CondVal { val: val2, of: 2 },
                    },
                ) => {
                    assert_eq!(cond1, cond2);

                    if *val1 == 0 && *val2 == 1 {
                        // swap then and else branches
                        SimpleBranch::If {
                            arg: cond1.clone(),
                            then_branch: self.graph[edge2.target()].name.clone(),
                            else_branch: self.graph[edge1.target()].name.clone(),
                        }
                    } else if *val1 == 1 && *val2 == 0 {
                        SimpleBranch::If {
                            arg: cond1.clone(),
                            then_branch: self.graph[edge1.target()].name.clone(),
                            else_branch: self.graph[edge2.target()].name.clone(),
                        }
                    } else {
                        panic!("Unexpected branch values");
                    }
                }
                _ => panic!(
                    "Invalid branch types {:?} and {:?}",
                    edge1.weight().op,
                    edge2.weight().op
                ),
            },
            _ => panic!("Too many outgoing edges"),
        }
    }
}

/// Get the underyling CFG corresponding to the function `func`.
///
/// The structure is reproduced exactly, aside from the addition of a single
/// exit node branched to from all return statements.
/// Generates a Cfg<Switch> because it returns a value in the annotation
pub(crate) fn function_to_cfg(func: &Function) -> SimpleCfgFunction {
    let mut builder = CfgBuilder::new(func);
    let mut block = Vec::new();
    let mut anns = Vec::new();
    let mut current = builder.cfg.entry;
    let mut had_branch = false;
    for inst in &func.instrs {
        match inst {
            Code::Label { label, pos } => {
                let next_block = builder.get_index(label);
                builder.finish_block(current, mem::take(&mut block), mem::take(&mut anns));
                builder.set_pos(next_block, pos.clone());
                if !had_branch {
                    builder.add_edge(
                        current,
                        next_block,
                        Branch {
                            op: BranchOp::Jmp,
                            pos: pos.clone(),
                        },
                    );
                }
                current = next_block;
                had_branch = false;
            }
            Code::Instruction(Instruction::Effect {
                args,
                funcs: _,
                labels,
                op: EffectOps::Branch,
                pos,
            }) => {
                had_branch = true;
                assert_eq!(labels.len(), 2, "unexpected format to branch instruction");
                assert_eq!(args.len(), 1, "unexpected format to branch instruction");
                let true_block = builder.get_index(&labels[0]);
                let false_block = builder.get_index(&labels[1]);
                let arg = &args[0];
                builder.add_edge(
                    current,
                    true_block,
                    Branch {
                        op: BranchOp::Cond {
                            arg: arg.into(),
                            val: true.into(),
                        },
                        pos: pos.clone(),
                    },
                );
                builder.add_edge(
                    current,
                    false_block,
                    Branch {
                        op: BranchOp::Cond {
                            arg: arg.into(),
                            val: false.into(),
                        },
                        pos: pos.clone(),
                    },
                );
            }
            Code::Instruction(Instruction::Effect {
                args: _,
                funcs: _,
                labels,
                op: EffectOps::Jump,
                pos,
            }) => {
                had_branch = true;
                assert_eq!(labels.len(), 1, "unexpected format to jump instruction");
                let dest_block = builder.get_index(&labels[0]);
                builder.add_edge(
                    current,
                    dest_block,
                    Branch {
                        op: BranchOp::Jmp,
                        pos: pos.clone(),
                    },
                );
            }
            Code::Instruction(Instruction::Effect {
                args,
                funcs: _,
                labels: _,
                op: EffectOps::Return,
                pos,
            }) => {
                had_branch = true;
                match args.as_slice() {
                    [] => {
                        builder.add_edge(
                            current,
                            builder.cfg.exit,
                            Branch {
                                op: BranchOp::Jmp,
                                pos: pos.clone(),
                            },
                        );
                    }
                    [arg] => {
                        anns.push(Annotation::AssignRet { src: arg.into() });
                        builder.add_edge(
                            current,
                            builder.cfg.exit,
                            Branch {
                                op: BranchOp::Jmp,
                                pos: pos.clone(),
                            },
                        );
                    }
                    _ => panic!("unexpected format to return instruction"),
                }
            }
            Code::Instruction(i) => block.push(i.clone()),
        }
    }
    builder.finish_block(current, block, anns);
    if !had_branch {
        builder.add_edge(
            current,
            builder.cfg.exit,
            Branch {
                op: BranchOp::Jmp,
                pos: None,
            },
        )
    }

    builder.cfg
}

struct CfgBuilder {
    cfg: SimpleCfgFunction,
    label_to_block: HashMap<String, NodeIndex>,
}

impl CfgBuilder {
    fn new(func: &Function) -> CfgBuilder {
        let mut graph = StableDiGraph::default();
        let entry = graph.add_node(BasicBlock::empty(BlockName::Entry));
        let exit = graph.add_node(BasicBlock::empty(BlockName::Exit));
        CfgBuilder {
            cfg: SimpleCfgFunction {
                args: func.args.clone(),
                graph,
                entry,
                exit,
                name: func.name.clone(),
                return_ty: func.return_type.clone(),
                _phantom: Simple,
            },
            label_to_block: HashMap::new(),
        }
    }

    fn get_index(&mut self, label: &str) -> NodeIndex {
        *self
            .label_to_block
            .entry(label.to_string())
            .or_insert_with(|| {
                self.cfg
                    .graph
                    .add_node(BasicBlock::empty(BlockName::Named(label.into())))
            })
    }

    fn finish_block(&mut self, index: NodeIndex, block: Vec<Instruction>, anns: Vec<Annotation>) {
        let BasicBlock { instrs, footer, .. } = self.cfg.graph.node_weight_mut(index).unwrap();
        debug_assert!(instrs.is_empty());
        debug_assert!(footer.is_empty());
        *instrs = block;
        *footer = anns;
    }

    fn set_pos(&mut self, index: NodeIndex, pos: Option<Position>) {
        self.cfg.graph.node_weight_mut(index).unwrap().pos = pos;
    }

    fn add_edge(&mut self, src: NodeIndex, dst: NodeIndex, branch: Branch) {
        self.cfg.graph.add_edge(src, dst, branch);
    }
}
