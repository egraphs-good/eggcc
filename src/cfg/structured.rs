use std::fmt::Display;

use super::BasicBlock;
use bril_rs::{Argument, Code, EffectOps, Function, Instruction, Program, Type};

#[derive(Debug, PartialEq, Clone)]
pub enum StructuredBlock {
    // Variable, then and else blocks
    Ite(String, Box<StructuredBlock>, Box<StructuredBlock>),
    Loop(Box<StructuredBlock>),
    Block(Box<StructuredBlock>),
    Sequence(Vec<StructuredBlock>),
    // how many layers of blocks / loops to break out of
    Break(usize),
    Return(Option<String>),
    Basic(Box<BasicBlock>),
}

#[derive(Debug)]
pub struct StructuredProgram {
    pub functions: Vec<StructuredFunction>,
}

impl Display for StructuredProgram {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut functions = self
            .functions
            .iter()
            .map(|f| format!("{}", f))
            .collect::<Vec<String>>();
        functions.sort();
        write!(f, "{}", functions.join("\n\n"))
    }
}

impl StructuredProgram {
    pub fn to_program(&self) -> Program {
        Program {
            functions: self
                .functions
                .clone()
                .into_iter()
                .map(|f| f.to_function())
                .collect(),
            imports: vec![],
        }
    }
}

#[derive(Debug, Clone)]
pub struct StructuredFunction {
    pub name: String,
    pub args: Vec<Argument>,
    pub return_ty: Option<Type>,
    pub block: StructuredBlock,
}

impl Display for StructuredFunction {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} {{\n{}\n}}", self.name, self.block)
    }
}

impl Display for StructuredBlock {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.display(0))
    }
}

impl StructuredBlock {
    fn display(&self, indent: usize) -> String {
        let indent = indent;
        let whitespace = " ".repeat(indent);
        match self {
            StructuredBlock::Ite(var, then, els) => format!(
                "{whitespace}if {}:\n{}\n{whitespace}else:\n{}",
                var,
                then.display(indent + 1),
                els.display(indent + 1)
            ),
            StructuredBlock::Loop(body) => {
                format!("{whitespace}while true:\n{}", body.display(indent + 1))
            }
            StructuredBlock::Block(body) => {
                format!("{}block:\n{}", " ".repeat(indent), body.display(indent + 1))
            }
            StructuredBlock::Sequence(blocks) => blocks
                .iter()
                .map(|b| b.display(indent))
                .collect::<Vec<String>>()
                .join("\n"),
            StructuredBlock::Break(n) => format!("{}break {}", " ".repeat(indent), n),
            StructuredBlock::Basic(block) => block
                .instrs
                .iter()
                .map(|i| format!("{}{i}", " ".repeat(indent)))
                .collect::<Vec<String>>()
                .join("\n"),
            StructuredBlock::Return(val) => {
                if let Some(val) = val {
                    format!("{whitespace}return {}", val)
                } else {
                    format!("{whitespace}return")
                }
            }
        }
    }
}

impl StructuredFunction {
    pub fn to_function(&self) -> Function {
        let mut builder = StructuredCfgBuilder {
            resulting_code: vec![],
            scopes: vec![],
            fresh_block_name_count: 0,
        };
        self.block.to_code(&mut builder);

        // TODO get rid of intermediate jumps
        Function {
            name: self.name.clone(),
            args: self.args.clone(),
            instrs: builder.resulting_code,
            pos: None,
            return_type: self.return_ty.clone(),
        }
    }
}

pub struct StructuredCfgBuilder {
    resulting_code: Vec<Code>,
    scopes: Vec<String>,
    fresh_block_name_count: usize,
}

impl StructuredCfgBuilder {
    fn fresh_block_name(&mut self) -> String {
        self.fresh_block_name_count += 1;
        format!("sblock___{}", self.fresh_block_name_count - 1)
    }

    /// find the name of the scope that we break to
    /// when breaking out of `num` layers of blocks
    /// we must break out of at least one block
    fn scope_break_to(&self, num: usize) -> String {
        assert!(num > 0);
        self.scopes[self.scopes.len() - num].clone()
    }
}

impl StructuredBlock {
    pub(crate) fn to_code(&self, builder: &mut StructuredCfgBuilder) {
        match self {
            StructuredBlock::Basic(block) => builder.resulting_code.extend(block.to_code()),
            StructuredBlock::Block(block) => {
                // add a label at the end of the block so
                // we can break out of it
                let end_of_block_label = builder.fresh_block_name();
                builder.scopes.push(end_of_block_label.clone());
                block.to_code(builder);
                builder.resulting_code.push(Code::Label {
                    label: end_of_block_label.clone(),
                    pos: None,
                });
                // pop the scope after we are done processing this block
                assert!(builder.scopes.pop().unwrap() == end_of_block_label);
            }
            StructuredBlock::Break(num) => {
                if *num != 0 {
                    builder
                        .resulting_code
                        .push(Code::Instruction(Instruction::Effect {
                            op: EffectOps::Jump,
                            args: vec![],
                            funcs: vec![],
                            labels: vec![builder.scope_break_to(*num)],
                            pos: None,
                        }));
                }
            }
            StructuredBlock::Ite(cond, then_block, else_block) => {
                let then_name = builder.fresh_block_name();
                let else_name = builder.fresh_block_name();
                let end_name = builder.fresh_block_name();
                builder
                    .resulting_code
                    .push(Code::Instruction(Instruction::Effect {
                        op: EffectOps::Branch,
                        args: vec![cond.to_string()],
                        funcs: vec![],
                        labels: vec![then_name.clone(), else_name.clone()],
                        pos: None,
                    }));

                // convert the then block
                builder.resulting_code.push(Code::Label {
                    label: then_name,
                    pos: None,
                });
                then_block.to_code(builder);
                // jump to the end
                builder
                    .resulting_code
                    .push(Code::Instruction(Instruction::Effect {
                        op: EffectOps::Jump,
                        args: vec![],
                        funcs: vec![],
                        labels: vec![end_name.clone()],
                        pos: None,
                    }));

                builder.resulting_code.push(Code::Label {
                    label: else_name,
                    pos: None,
                });

                else_block.to_code(builder);

                // add a label at the end of the if statement
                builder.resulting_code.push(Code::Label {
                    label: end_name,
                    pos: None,
                });
            }
            StructuredBlock::Loop(block) => {
                // we need to be able to loop back to the start
                let loop_start_name = builder.fresh_block_name();
                builder.resulting_code.push(Code::Label {
                    label: loop_start_name.clone(),
                    pos: None,
                });

                // loops can be broken out of
                let loop_end_name = builder.fresh_block_name();
                builder.scopes.push(loop_end_name.clone());
                block.to_code(builder);

                // jump back to the start of the loop if you get to the end
                builder
                    .resulting_code
                    .push(Code::Instruction(Instruction::Effect {
                        op: EffectOps::Jump,
                        args: vec![],
                        funcs: vec![],
                        labels: vec![loop_start_name],
                        pos: None,
                    }));
                builder.resulting_code.push(Code::Label {
                    label: loop_end_name.clone(),
                    pos: None,
                });
                assert!(builder.scopes.pop().unwrap() == loop_end_name);
            }
            StructuredBlock::Sequence(blocks) => {
                for block in blocks {
                    block.to_code(builder);
                }
            }
            StructuredBlock::Return(val) => {
                let args = match val {
                    Some(v) => vec![v.clone()],
                    None => vec![],
                };
                builder
                    .resulting_code
                    .push(Code::Instruction(Instruction::Effect {
                        op: EffectOps::Return,
                        args,
                        funcs: vec![],
                        labels: vec![],
                        pos: None,
                    }));
            }
        }
    }
}
