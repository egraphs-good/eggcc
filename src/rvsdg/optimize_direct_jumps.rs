//! This module takes as input a [`SimpleCfgProgram`] and gets rid of any
//! direct jumps between blocks.
//! The block that is jumped to must have only one predecessor.
//! This is used by `to_cfg` to clean up
//! the output.

use hashbrown::HashMap;
use petgraph::{
    graph::EdgeIndex,
    stable_graph::{NodeIndex, StableDiGraph, StableGraph},
    visit::{DfsPostOrder, EdgeRef},
    Direction,
};

use crate::cfg::{BasicBlock, Branch, Simple, SimpleCfgFunction, SimpleCfgProgram};

#[cfg(test)]
use crate::Optimizer;

impl SimpleCfgFunction {
    pub fn optimize_jumps(&self) -> Self {
        self.fuse_down().collapse_empty_blocks()
    }

    /// Find cases where a block jumps directly to another block A -> B where B
    /// A has only one outgoing edge and B has one incoming edge
    /// Turn it into one block AB
    fn fuse_down(&self) -> SimpleCfgFunction {
        let mut resulting_graph: StableGraph<BasicBlock, Branch> = StableDiGraph::new();

        // a map from nodes in the old graph to nodes in the
        // new graph
        // if a node was fused into another node,
        // it points to the new, fused node
        let mut node_mapping: HashMap<NodeIndex, NodeIndex> = HashMap::new();

        // we use a dfs post order
        // so dependencies are visited before parents
        // This ensures that `node_mapping[&next]` succeeds.
        let mut dfs = DfsPostOrder::new(&self.graph, self.entry);

        let mut edges_to_add = vec![];

        // copy the graph without the edges
        // also choose which nodes get fused to which
        // by re-assigning in the node map
        while let Some(node) = dfs.next(&self.graph) {
            let outgoing_from_node = self
                .graph
                .edges_directed(node, Direction::Outgoing)
                .collect::<Vec<_>>();
            let target = if let &[single_edge] = outgoing_from_node.as_slice() {
                let target = single_edge.target();
                let incoming_to_next = self
                    .graph
                    .edges_directed(target, Direction::Incoming)
                    .count();
                if incoming_to_next == 1 && target != node {
                    Some(target)
                } else {
                    None
                }
            } else {
                None
            };
            // single outgoing edge
            if let Some(next) = target {
                let new_target = node_mapping[&next];

                // this node will be mapped to the previous
                node_mapping.insert(node, new_target);

                // add instructions to the beginning of the next node
                let mut new_instrs = self.graph[node].instrs.to_vec();
                let mut new_footer = self.graph[node].footer.to_vec();
                new_instrs.extend(resulting_graph[new_target].instrs.to_vec());
                new_footer.extend(resulting_graph[new_target].footer.to_vec());

                resulting_graph[new_target].instrs = new_instrs;
                resulting_graph[new_target].footer = new_footer;
            } else {
                // add the node
                let new_node = resulting_graph.add_node(self.graph[node].clone());
                node_mapping.insert(node, new_node);

                edges_to_add.extend(self.graph.edges_directed(node, Direction::Outgoing));
            }
        }

        for edge in edges_to_add {
            let source = node_mapping[&edge.source()];
            let target = node_mapping[&edge.target()];
            resulting_graph.add_edge(source, target, edge.weight().clone());
        }

        SimpleCfgFunction {
            name: self.name.clone(),
            args: self.args.clone(),
            graph: resulting_graph,
            entry: node_mapping[&self.entry],
            exit: node_mapping[&self.exit],
            _phantom: Simple,
            return_ty: self.return_ty.clone(),
        }
    }

    // this function looks for all CFG fragments of the form
    // source node --(source edge)-> empty node --(target edge)-> target node
    // (where there is only one target edge and the empty node has no instructions)
    // and changes the source edge to point to the target node
    fn collapse_empty_blocks(mut self) -> SimpleCfgFunction {
        // loop over every non-empty block, and look at its parents
        for block in self.graph.node_indices().collect::<Vec<_>>() {
            // find parents
            let parents: Vec<NodeIndex> = self
                .graph
                .edges_directed(block, Direction::Incoming)
                .map(|edge| edge.source())
                .collect();

            for parent in &parents {
                self.collapse_empty_block(*parent);
            }
        }

        self
    }

    /// Detect the case where source -> empty -> target
    /// The empty block should have a single incoming and single outgoing edge
    fn get_single_in_single_out(&self, parent_block: NodeIndex) -> Option<(EdgeIndex, EdgeIndex)> {
        if !self.graph[parent_block].instrs.is_empty()
            || !self.graph[parent_block].footer.is_empty()
        {
            return None;
        }

        let outgoing = self
            .graph
            .edges_directed(parent_block, Direction::Outgoing)
            .collect::<Vec<_>>();
        let incoming = self
            .graph
            .edges_directed(parent_block, Direction::Incoming)
            .collect::<Vec<_>>();
        if let ([source_edge], [outgoing]) = (incoming.as_slice(), outgoing.as_slice()) {
            Some((source_edge.id(), outgoing.id()))
        } else {
            None
        }
    }

    /// Detect the case when source -> empty -> target
    /// and collapse it to source -> target
    fn collapse_empty_block(&mut self, parent_block: NodeIndex) {
        if let Some((source_edge, outgoing)) = self.get_single_in_single_out(parent_block) {
            let weight = self.graph.edge_weight(source_edge).unwrap().clone();
            let (source, empty) = self.graph.edge_endpoints(source_edge).unwrap();
            let (empty_, target) = self.graph.edge_endpoints(outgoing).unwrap();
            assert_eq!(empty, empty_);

            self.graph.remove_edge(source_edge);
            self.graph.add_edge(source, target, weight);
        }
    }
}

impl SimpleCfgProgram {
    pub fn optimize_jumps(&self) -> Self {
        SimpleCfgProgram {
            functions: self
                .functions
                .iter()
                .map(|f| {
                    // NB: We could avoid this copy by having `optimize_jumps` take `self` by value.
                    let mut res = f.optimize_jumps();
                    res.simplify_branches();
                    res
                })
                .collect(),
        }
    }
}

#[test]
fn single_node() {
    // TODO these imports are very bad
    use crate::cfg::BlockName;
    use crate::test_util::*;

    let mut graph = StableDiGraph::new();
    let node = graph.add_node(BasicBlock {
        name: BlockName::Entry,
        instrs: vec![],
        footer: vec![],
        pos: None,
    });
    let input_cfg = SimpleCfgFunction {
        args: vec![],
        graph,
        entry: node,
        exit: node,
        name: "test".to_string(),
        _phantom: Simple,
        return_ty: None,
    };

    cfg_test_equiv!(input_cfg.optimize_jumps(), []);
}

#[test]
fn loops_to_self() {
    // TODO you have to import both of these
    //  is there a way to package up some macros all together?
    use crate::cfg::BlockName;
    use crate::test_util::*;
    let mut graph = StableDiGraph::new();
    let node = graph.add_node(BasicBlock {
        name: BlockName::Entry,
        instrs: vec![],
        footer: vec![],
        pos: None,
    });
    graph.add_edge(
        node,
        node,
        Branch {
            op: crate::cfg::BranchOp::Jmp,
            pos: None,
        },
    );

    let input_cfg = SimpleCfgFunction {
        args: vec![],
        graph,
        entry: node,
        exit: node,
        name: "test".to_string(),
        _phantom: Simple,
        return_ty: None,
    };

    cfg_test_equiv!(input_cfg.optimize_jumps(), [ENTRY = (Jmp)=> ENTRY,]);
}

#[test]
fn add_block_ind_test() {
    let prog = include_str!("../../tests/passing/small/add_block_indirection.bril");
    let cfg = Optimizer::program_to_cfg(&Optimizer::parse_bril(prog).unwrap());
    insta::assert_snapshot!(cfg.optimize_jumps().to_bril().to_string());
}
