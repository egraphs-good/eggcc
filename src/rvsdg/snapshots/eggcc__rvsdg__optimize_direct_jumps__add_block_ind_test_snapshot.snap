---
source: src/rvsdg/optimize_direct_jumps.rs
expression: cfg.optimize_jumps().to_bril().to_string()
---
@main {
.exit___:
  v0: int = const 1;
  v1: int = const 2;
  v2: int = add v0 v1;
  print v2;
  ret;
}
